src/Lib.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Lib
+    (
+     ascii_art_str,
+     basic_str,
+     basic_const_zero,
+     BasicExternal,
+    ) where
+
+import Foreign.C.Types
+import Foreign.Ptr
+import Foreign.C.String
+import Foreign.Storable
+import Foreign.Marshal.Array
+import Foreign.ForeignPtr
+import Control.Applicative
+
+
+data BasicStruct = BasicStruct {
+    data_ptr :: Ptr ()
+}
+
+instance Storable BasicStruct where
+    alignment _ = 8
+    sizeOf _ = sizeOf nullPtr
+    peek basic_ptr = BasicStruct <$> peekByteOff basic_ptr 0
+    poke basic_ptr BasicStruct{..} = pokeByteOff basic_ptr 0 data_ptr
+
+
+type BasicExternal = ForeignPtr BasicStruct
+-- !an array of size 1 of type `BasicStruct`
+type BasicInternal = Ptr BasicStruct
+-- exported functions
+
+
+basic_const_zero :: IO BasicExternal
+basic_const_zero = do
+    basic <- create_basic
+    withForeignPtr basic basic_const_zero_raw
```

Comment:
If you have a function that operates on a `Ptr` (which is the "right" way to operate on a FFI pointer), the [`withForeignPtr'](http://hackage.haskell.org/package/base-4.8.2.0/docs/Foreign-ForeignPtr.html#v:withForeignPtr) function "promotes" the function to be able to act on a `ForeignPtr`.

```
withForeignPtr :: ForeignPtr a -> (Ptr a -> IO b) -> IO b
                           ^-| foreign     ^-| external ^-| output
                             | pointer       | world action

```